ScriptEngine: Expose ScriptEngine as OSGI-Bundle, ScriptExtension, GlobalScope optimization

smerschjohann opened the pull request:
This PR will allow other bundles to use and instantiate ScriptEngines.

It also contains a GlobalScope optimization, which is required to circumvent some OSGI related issues concerning the current implementation of class loading for the Nashorn-based solution. It will use "variable = variable.static" instead of "variable = Java.type(...)". In addition the Scope will be initialized only one time and injected as GLOBAL_SCOPE instead of ENGINE_SCOPE.

Futhermore this PR introduces a ScriptExtensionProvider. This allows OSGI-Bundles to provide functionality which is not needed at all times. The given features can be requested from a script and are not imported in all ScriptEngine instances. It furthermore allows to keep track of the loaded instances based on a ScriptEngineId. If the ScriptEngine gets unloaded, the ScriptExtensionProvider will be notified about this and can unload or undo changes it has done to the runtime.

kgoderis commented:
Hi Johann, I am going through a lot of PR related to Automation, including this one. I am in the process of rewriting the script bundles so that they make use of the Automation API. In the process I have done away with the ScriptEngine (as it was), and rather built something that is JSR223 compliant, and that is also OSGi'fied (see https://github.com/eclipse/smarthome/issues/900). I think a lot of the changes/extensions you proposed could be integrated.

smerschjohann commented:
As I don't know what you actually did, it is quiet hard to comment on it ;)
Of course I want to see my changes integrated into smarthome, if not in exactly this form, as I have already seen some parts that could be changed for better code reuse (not this PR), than in a quiet similiar way.
I'm currently waiting for a general feedback before looking into it again.

I don't know what you mean by "JSR223 compliant" as these things are JSR223 compliant as well. Your addition concerning classloader replacements for OSGI-bundles would be a nice addition to this one.

"So that they make use of the automation API": one of my PRs do exactly that

lewie commented:
@smerschjohann,
yea, your bundle for OH1.8 is really great to work with.
All my scripts run stable as a rock. 
And they are much more complex, extremely fast, many times better structured and more flexible than any ever with xText. For a Year now! And I use openHAB and xText since 2011.



@kgoderis, 
I included nashorn for JS for the JSR223 binding.
OSGi frameworks like felix or eqinox define the class-loader limits for running bundles.
There are mainly three models:
- *boot* for the boot class loader (most default)
- *app* for the application class loader
- *ext* for the extension class loader

For using nashorn and other Script-engines with OSGI [*ext* is needed](https://github.com/openhab/openhab/wiki/Jsr223-Script-Engine). Its straddling the allowed range of classes for the bundle in OSGI.
In/for Java9 everything is modularized therefore *ext* will be standard class-loader.
More information about you can find if you google java-fx (uses nashorn too) or for example: http://spring.io/blog/2009/01/19/exposing-the-boot-classpath-in-osgi/ ...

There is no particular problem of OSGi, to allow load classes from the outside, it is a problem of the principle of modularity.

Can not see it, but for me it would be interesting, if there is another way to solve this Problem. By a OSGI service, dynamically release directories ...

kgoderis commented:
@lewie I don't visit OH1's wiki that often, so - shame on me - I was not aware of what you did. Now, to be specific, what I am currently doing is migrating the Xtext based script-engine, in ESH, over to the ESH.Automation framework, thereby exposing .rules and .scripts as ScriptActions (see https://github.com/eclipse/smarthome/tree/master/bundles/automation/org.eclipse.smarthome.automation.module.script). In essence, what is does is take the DSL .script files, and render them as ESH.Automation Rules. Now, the ScriptAction, as it is defined today, builds on top of javax.script.* classes, which is using JSR223 for loading the ScriptEngines. I merely (https://github.com/eclipse/smarthome/issues/900) applied the solution proposed in https://devnotesblog.wordpress.com/2011/09/07/scripting-using-jsr-223-in-an-osgi-environment/ to solve the class loading issue. In the meantime I removed the exposure as OSGi service (not really needed in fact). So, it stills relies on the current Xtext-based DSL for scripts, but adding any other script language would be rather easy, and we could even imagine embedding one script language into another one and so forth

@smerschjohann I definitely are going to read up on what you already PR'd. I simply did all of this as an intermediate step for the CalDAV servlet (e.g. harmonise the rule engines so that I only have to deal with Automation classes). Do not have the intention to re-invent the wheel here.

What we however could reflect upon, based on javax.script, is what POJO's we would like to expose to the ScriptEngines (at least, do that in a harmonised manner) so that we more or less have the same features in each Engine. And secondly, how we define and use the Context and Scopes (e.g. Context set at the level of the Script, the ScriptEngine, or ScriptEngineManager (e.g. all ScriptEngines))

kgoderis commented:
@smerschjohann This is an interesting concept, but wonder if we could put this on top of what javax.script is already providing? Unknowingly of this PR, I kinda did away with ScriptEngine, and build on top of javax.script.ScriptEngine since that was hinted for in the ESH.Automation codebase. It is more less answers the question in the previous comment on how to jockey context and scopes between ScriptEngines

lewie commented:
@smerschjohann, wow, this is really exciting!
[They did the trick to solve this annoying OSGI classloader facts.](http://svn.apache.org/repos/asf/felix/trunk/mishell/src/main/java/org/apache/felix/mishell/OSGiScriptEngineManager.java)

I am very eager to test the new script engines!

Thanks for the clarification.
Helmut

smerschjohann commented:
@kgoderis it will very likely not work with your proposed way without even looking at it ;)

The Script-Extension "framework" this PR introduces is needed for the extensibility in an easy way in the script engines. It further allows to have a context in the extensions which enables the extensions to behave in a certain way if the scripts are reloaded, loaded and unloaded.

If your main goal is to modulify the xtend languages, you could simply use the ScriptManager introduced in this PR. It will give you a javax.script.ScriptEngine which has some default scopes injected into it (like the ability to load ScriptExtensions)

smerschjohann commented:
@lewie Yes, I have already seen that, but did not have the time to look into it. As I thought, @kgoderis was refering exactly that kind of classloader manipulations ;). The way I change the types in this PR (variable = variable.static for nashorn) makes this kind of optional in most cases, but nontheless this can be a very useful addition to this PR.

Just for information, the changes related to script and rule support are split into multiple bundles and therefore into multiple PRs. They are:

- This PR
- The ability to dynamically change ModuleHandlers (this is needed for the RuleSupport extension): https://github.com/eclipse/smarthome/pull/805
- The ability to load scripts from a directory and from bundles (@kgoderis you might want to look into this one) : https://github.com/eclipse/smarthome/pull/806
- The ScriptExtension: RuleSupport to enable writing rules and modules in scripts: https://github.com/eclipse/smarthome/pull/813

kgoderis commented:
Ok - For now I tried to stay away "as far as possible" from the Xtext specific bits. the objective was not to change things too dramatically so that the proper working of the changes can be tested. I also added a  SimpleScriptContext derived class with additional scopes. It still works with the ScopeProvider classes Kai added. I need to revisit your PRs to fully understand things.

I propose to PR what I have been cooking, and then see how we can glue everything together?

smerschjohann commented:
You can look into my code, as it is already present as PR. I cannot in yours, so it's up to you. I don't know what you've added and therefore what is missing in my PR. With the sum of my PRs you can write rules and modules in scripts, so I don't really see what is missing here, except the OSGI classloader addition you already mentioned. But of course, you can commit yours and link to your git-branch here.

But I think you are doing the effort a second time in a lot of places as these PRs already provide the scriptable automation feature with the JSR223 extension of openhab1 in mind. The Script-Extension support is a thing needed or at least beneficial for that. Also, my additions keep in mind that scripts can be loaded and unloaded which leads to rule/module removal etc. (a thing a script-extension can deal with) and allow scripts to be loaded from directory. It might be a better idea to tell me what is missing here or fork my branch and simply add what you think is missing. I will look into it then.

Nontheless, there might be areas that needs to be fixed in general as @kaikreuzer has not looked into this code until now.

kgoderis commented:
Well, it is actually very complementary code. You added the possibility to write Rules from scripts, and I transformed the ScriptEngine so that it runs on Rules internally, with no change to the Script inputs. We just need to connect the two I think.

smerschjohann commented:
Do you mean you transformed the Xtend-Engine in combination with its RuleEngine and transformed the defined Rules-Format into a format where the "then" part is defined by a JSR223 compliant language? Or what do you mean by you "transformed the ScriptEngine so that it runs on Rules internally"?

Or have you even gone one step further and have written a new "script language" for Jsr223?

For me it would be much faster and easier to follow, if you could commit your code, even if it is unfinished ;)

kgoderis commented:
Johann, no, the rules format is still the same, and there is no change in script language (e.g. Xtext). I have made a preview PR that you can find here https://github.com/kgoderis/smarthome/pull/1 . It is in comparison to a branch that already has https://github.com/eclipse/smarthome/pull/869 integrated.

kgoderis commented:
@smerschjohann do you already have a concrete implementation of the the ScriptExtensionProvider? Just to understand, these are meant to be stand-alone implementations that go with a given type of script engine, or will it be a mechanism for the RuleRegistry/Engine to inject scope into ScriptEngines? (e.g. RR or RE becoming a ScriptExtensionProvider). Just side note, trying to understand why you added a ScriptExtensionManager, e.g. have ScriptEngineProvider manage and consume the ScriptExtensionProviders directly?

My feeling is that since the nashorn js engine is "pre-defined" in the JVM class path, you did not run into the class loading problem. That one pops up when you add a custom script engine, e.g. the Xtext one. We could try to integrate ScriptEngineProvider and ScriptExtensionProvider with the OSGI* classes from the Apache framework (see my PR) to generalise it for all script engine types

kgoderis commented:
ok - https://github.com/smerschjohann/smarthome/blob/feature/script_ext_rulesupport/bundles/automation/org.eclipse.smarthome.automation.module.script.extensions.rulesupport/src/main/java/org/eclipse/smarthome/automation/module/script/extensions/rulesupport/internal/LoaderScriptExtension.java answers my previous question

lewie commented:
@kgoderis, nashorn can only "see" its parent OSGI container classes, only classes imported in OSGI container (example jsr223-Binding).

smerschjohann commented:
> Just side note, trying to understand why you added a ScriptExtensionManager, e.g. have ScriptEngineProvider manage and consume the ScriptExtensionProviders directly?

It is meant for optional addons. You would import extensions that you want to use in your scripts, this helps to keep the scope free from unneeded variables. It can as well help in my compatbility approach, I would add another extension which provides compatibility for jsr223 of openhab1. 

> We could try to integrate ScriptEngineProvider and ScriptExtensionProvider with the OSGI* classes from the Apache framework (see my PR) to generalise it for all script engine types

Yes, I think this is a good idea, even if it is not (really) needed for the automation API.

lewie commented:
@smerschjohann, @kgoderis, 
any News about alternative scripting in OH2? 
I really need javascript support in OH2!
Is there currently any working/testable code in OH2 with which I can perform javascript?
If not I "simply" would port jsr223 Binding to OH2?! 
JSR223:
May have you tips on how to fix the class loader problem under Karaf? Under eqinox I simply had to start java with "-Dorg.osgi.framework.bundle.parent = ext". Does this work for karaf too?

kaikreuzer commented:
The new rule engine is not yet available on OH2, this is still on my to-do list and it is a prerequisite for the work @smerschjohann has done here.

@danchom & @marinmitev Did you already review this (and the other related) PR and have any comments/feedback?

smerschjohann commented:
Yes and no. My PRs in a whole allow Scripting by JavaScript. Until now they
are not reviewed.

For compatibility with oh1: you would need to provide the triggertypes as
there is no equivalent defined for some in the new rule engine.
Am 09.03.2016 8:49 vorm. schrieb "Helmut Lehmeyer" <notifications@github.com
>:

> @smerschjohann <https://github.com/smerschjohann>, @kgoderis
> <https://github.com/kgoderis>,
> any News about alternative scripting in OH2?
> I really need javascript support in OH2!
> Is there currently any working/testable code in OH2 with which I can
> perform javascript?
> If not I "simply" would port jsr223 Binding to OH2?!
> JSR223:
> May have you tips on how to fix the class loader problem under Karaf?
> Under eqinox I simply had to start java with
> "-Dorg.osgi.framework.bundle.parent = ext". Does this work for karaf too?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/eclipse/smarthome/pull/803#issuecomment-194166076>.
>

lewie commented:
@smerschjohann,
do you mean [this](https://github.com/smerschjohann/smarthome/commit/cd3c635dc37695d22c2ebf125e26031f9d472aff)? Did not realize the skills of your PR. I'll try to expand it and test. Thanks. 

@kaikreuzer,
Thank you very much, for pushing the new rule engine! :-)

danchom commented:
Unfortunately I'm not in depth in script handler implementation, and I don't understand this PR. I'm going to learn about it in near feature. 
Аt first look these PRs add third party dependences or import internal packages, which I think they must be workarounded:
1) #803 
 org/eclipse/smarthome/automation/module/script/ScriptEngineProvider.java
    +import com.google.common.base.Joiner;
    +import org.eclipse.smarthome.automation.module.script.internal.ScriptExtensionManager;
    +import org.eclipse.smarthome.automation.module.script.internal.handler.AbstractScriptModuleHandler;
2) #806 
org.eclipse.smarthome.automation.module.script.loader/META-INF/MANIFEST.MF
    +Import-Package: com.google.common.base;version="10.0.1",
 + com.google.common.collect,
 + org.apache.commons.collections.keyvalue,
 + org.apache.commons.io,
 + org.apache.commons.lang,

Kai ,could you please review it?

kgoderis commented:
@kaikreuzer @danchom @lewie I have a PR ready with a script engine that is built on top of the Automation buildings blocks. I have not pushed it yet as it depends on the new scheduler infrastructure, and it is to be debated with respect to this PR as well

kaikreuzer commented:
@smerschjohann, @kgoderis, @lewie: My sincere apologies to not coming back to this any earlier - I very much appreciate your patience and hope that you are still willing to continue on this!

The good news is: I have now finally made the new rule engine available as an experimental feature in openHAB! To install it, do the following:
- get the latest distro from https://openhab.ci.cloudbees.com/job/openHAB-Distribution/
- Go to the Paper UI -> Configuration -> Services -> Extension Management
- Enable "Access Remote Repositories" and "Include Experimental Extensions" and save
- Go to Extensions -> Misc and install "Rule Engine (Experimental)"
- If you now refresh the Paper UI, you should see a "Rules" menu entry appearing

Within the openHAB IDE, you can also simply add the ESH automation bundles to the workspace and add them to the launch config and you can directly use them from openHAB being launched in the IDE.

Now the question to you on how to proceed on the PRs: Shall @smerschjohann update this PR first and make it mergeable again? Or do you @kgoderis already have code ready that is based on @smerschjohann PR and we should go for yours?

kgoderis commented:
@kaikreuzer Johann's code and my bits are not based on each other. I have not yet updated my code though.

lewie commented:
Only for info: Current version (yesterday) I get this error when installing Rule Engine (Experimental)

```
14:22:27.341 [INFO ] [.dashboard.internal.DashboardService] - Started dashboard at /start
14:22:27.413 [ERROR] [rg.eclipse.smarthome.automation.core] - FrameworkEvent ERROR - org.eclipse.smarthome.automation.core
org.osgi.framework.BundleException: Exception in org.eclipse.smarthome.automation.core.internal.Activator.start() of bundle org.eclipse.smarthome.automation.core.
        at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:792)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:721)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:936)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:319)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.container.Module.doStart(Module.java:571)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.container.Module.start(Module.java:439)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1582)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1562)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1533)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1476)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
Caused by: java.lang.NoSuchMethodError: org.eclipse.smarthome.core.common.registry.AbstractRegistry.setManagedProvider(Lorg/eclipse/smarthome/core/common/registry/ManagedProvider;)V
        at org.eclipse.smarthome.automation.core.internal.RuleRegistryImpl.setManagedProvider(RuleRegistryImpl.java:57)
        at org.eclipse.smarthome.automation.core.internal.Activator$1.addingService(Activator.java:113)
        at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:941)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:870)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.AbstractTracked.trackInitial(AbstractTracked.java:183)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:318)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:261)[org.osgi.core-6.0.0.jar:]
        at org.eclipse.smarthome.automation.core.internal.Activator.start(Activator.java:161)
        at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:771)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at java.security.AccessController.doPrivileged(Native Method)[:1.8.0_74]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:764)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        ... 12 more
```

maggu2810 commented:
@lewie Is your error caused by a runtime based on the branch tracked by this PR? Or how it is related to this PR?

MichaelWeller commented:
I spend several hours to follow up what’s going on with rules and scripting in OH2.
**I had some success and got a simple rule running:  Switch on a light by item name from javascript.**

BUT is extremely hard for me to find and understand the information pieces. 

Maybe it’s a good time to start some documentation on scripting, even if something are not final today. For me (and maybe for some of others too) scripting is one of most imported features of OH.

My personal biggest issues is find some information for the current api / class binding I can use in a script. Even going through the source code doesn’t help me, because I obviously didn’t grab the file that implements this binding. Maybe someone could give me some direction where to look, or even some documentation ;-)

There is a second import point: I should to be able to “include” some simple js-libraries.
Even for this topic I can’t get the essence of the discussion above…

Would be great to post some more success here ….

lewie commented:
Oh, sorry did not know about another Branch. Took yesterday [last stable Build from cloudbees](https://openhab.ci.cloudbees.com/job/openHAB-Distribution/lastStableBuild/artifact/distributions/openhab-online/target/openhab-online-2.0.0-SNAPSHOT.zip).

maggu2810 commented:
Hi guys, please differ between comments to this PR and other issues about the rule engine.
This PR is not part of the master branch so I assume you have not used this one.
Comments about the current state of the rule, script, ... features should be placed in new / existing issues or in the forum.
If you use this branch or do want to comment the changes this branch will made, then this is the correct place.
All other comments will blow up this PR and does not help at all.

If you have moved your comments to another issue or the forum it would be great if you could remove the miss-placed one here (I will remove mine after that, too).

Thanks for your patience.

lewie commented:
Ok, deleted my comments.

danchom commented:
As I said I'm not in deep in script engines and as the PR is related to script module only, it is ok  for me to be merged. I'll just ask for, if it is possible, third party dependencies  (i.e. com.google.common.base.Joiner) to be escaped.

smerschjohann commented:
I will look into it in some weeks when I have a bit more time. I think most imports can be dropped in one or the other way. But we'll see.

smerschjohann commented:
ok, I've merged and updated the three remaining PRs (https://github.com/eclipse/smarthome/pull/813, https://github.com/eclipse/smarthome/pull/806 and this one) into a dev branch: https://github.com/smerschjohann/smarthome/tree/automation-scriptable

I've added a working test script at: distribution/smarthome/conf/scripts/test.py

Basically, everything is functional. If you want, I can split the dev branch into the PRs again, but I just wanted you to have a look at it first ;)

I removed every dependency to third party libraries as well.

kaikreuzer commented:
@smerschjohann That sounds excellent. I hope to be able to look at it in the next few days!
Would it be possible that you create a PR from this branch as this makes it easier to add comments?
We could then close #813, #806 and this and continue any further discussion on the new PR.

@kgoderis Any chance that you also have a look at it soon?

smerschjohann commented:
@kaikreuzer I've just opened the new PR.

This PR is superseeded by https://github.com/eclipse/smarthome/issues/1783
